pilot/pkg/serviceregistry/kube/controller/filter/discoverynamespaces.go

```diff
@@ -0,0 +1,185 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filter
+
+import (
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	listerv1 "k8s.io/client-go/listers/core/v1"
+
+	"istio.io/pkg/log"
+)
+
+// DiscoveryNamespacesFilter tracks the set of namespaces selected for discovery, which are updated by the discovery namespace controller.
+// It exposes a filter function used for filtering out objects that don't reside in namespaces selected for discovery.
+type DiscoveryNamespacesFilter interface {
+	// return true if the input object resides in a namespace selected for discovery
+	Filter(obj interface{}) bool
+	// invoked when meshConfig's discoverySelectors change, returns any newly selected namespaces and deselected namespaces
+	SelectorsChanged(discoverySelectors []*metav1.LabelSelector) (selectedNamespaces []string, deselectedNamespaces []string)
+	// return true if the created namespace is selected for discovery
+	NamespaceCreated(ns metav1.ObjectMeta) (membershipChanged bool)
+	// membershipChanged will be true if the updated namespace is newly selected or deselected for discovery
+	NamespaceUpdated(oldNs, newNs metav1.ObjectMeta) (membershipChanged bool, namespaceAdded bool)
+	// return true if the deleted namespace was selected for discovery
+	NamespaceDeleted(ns metav1.ObjectMeta) (membershipChanged bool)
+}
+
+type discoveryNamespacesFilter struct {
+	lock                sync.RWMutex
```

Comment:
this is a RWMutex but we are always using .Lock() when we are only reading. We should switch to RLock